pkg/healthmonitor/health_monitor.go

```diff
@@ -0,0 +1,338 @@
+				newHealthyTargets = append(newHealthyTargets, svrErrTuple.target)
+			}
+		}
+	}
+
+	newUnhealthyTargetsSet := sets.NewString(newUnhealthyTargets...)
+	newHealthyTargetsSet := sets.NewString(newHealthyTargets...)
+	notifyListeners := false
+
+	// detect unhealthy targets
+	previouslyUnhealthyTargetsSet := sets.NewString(sm.unhealthyTargets...)
+	currentlyUnhealthyTargetsSet := previouslyUnhealthyTargetsSet.Union(newUnhealthyTargetsSet)
+	currentlyUnhealthyTargetsSet.Delete(newHealthyTargetsSet.List()...)
+	if !currentlyUnhealthyTargetsSet.Equal(previouslyUnhealthyTargetsSet) {
+		sm.unhealthyTargets = currentlyUnhealthyTargetsSet.List()
+		klog.V(2).Infof("observed the following unhealthy targets %v", sm.unhealthyTargets)
+		logUnhealthyTargets(sm.unhealthyTargets, currentHealthCheckProbes)
+
+		exportedUnhealthyTargets := make([]string, len(sm.unhealthyTargets))
+		for index, unhealthyTarget := range sm.unhealthyTargets {
+			exportedUnhealthyTargets[index] = unhealthyTarget
+		}
+		sm.exportedUnhealthyTargets.Store(exportedUnhealthyTargets)
+		notifyListeners = true
+	}
+
+	// detect healthy targets
+	previouslyHealthyTargetsSet := sets.NewString(sm.healthyTargets...)
+	currentlyHealthyTargetsSet := previouslyHealthyTargetsSet.Union(newHealthyTargetsSet)
+	currentlyHealthyTargetsSet.Delete(newUnhealthyTargetsSet.List()...)
+	if !currentlyHealthyTargetsSet.Equal(previouslyHealthyTargetsSet) {
+		sm.healthyTargets = currentlyHealthyTargetsSet.List()
+		klog.V(2).Infof("observed the following healthy targets %v", sm.healthyTargets)
+
+		exportedHealthyTargets := make([]string, len(sm.healthyTargets))
+		for index, healthyTarget := range sm.healthyTargets {
+			exportedHealthyTargets[index] = healthyTarget
+		}
+		sm.exportedHealthyTargets.Store(exportedHealthyTargets)
+		notifyListeners = true
+	}
+
+	if notifyListeners {
+		// notify listeners about the new healthy/unhealthy targets
+		for _, listener := range sm.listeners {
+			listener.Enqueue()
+		}
+	}
+}
+
+func (sm *HealthMonitor) healthCheckSingleTarget(target string) error {
+	// TODO: make the protocol, port and the path configurable
+	url := fmt.Sprintf("https://%s/%s", target, "readyz")
+	newReq, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return err
+	}
+
+	resp, err := sm.client.Do(newReq)
+	if err != nil {
+		return err
+	}
+	resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("bad status from %v: %v, expected HTTP 200", url, resp.StatusCode)
+	}
+
+	return err
+}
+
+func createHealthCheckHTTPClient(responseTimeout time.Duration, restConfig *rest.Config) (*http.Client, error) {
+	transportConfig, err := restConfig.TransportConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	tlsConfig, err := transport.TLSConfigFor(transportConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	client := &http.Client{
+		Transport: utilnet.SetTransportDefaults(&http.Transport{
+			TLSClientConfig: tlsConfig,
+		}),
+		Timeout: responseTimeout,
```

Comment:
probably ignore the other comment, but make sure this timeout really is the maximum timeout the request will take... including DNS resolving

Reply:
there is `probeResponseTimeout` parameter for specifying the maximum time limit for requests made by the HTTP client